### PR TITLE
Accomodate provider spec

### PR DESCRIPTION
**Pull request by ingvagabund**
Built on top of https://github.com/openshift/cluster-api-provider-libvirt/pull/103

Due to the following breaking change https://github.com/kubernetes-sigs/cluster-api/pull/548 we need to support both `.providerConfig` and `.providerSpec` fields until all consumers of the libvirt actuator migrate to new `.providerSpec` field.

We could just keep the `.providerConfig` and make sure it gets de-serialized into `ProviderSpec` struct. Though, folks could get confused by that and we should stay with the upstream name convention.

Once merged, we can rename the following:
- `providerConfig`
- `ProviderConfig`
- `DecodeProviderConfig`
- `EncodeProviderConfig`
- `LibvirtMachineProviderConfig`

to corresponding:
- `providerSpec`
- `ProviderSpec`
- `DecodeProviderSpec`
- `EncodeProviderSpec`
- `LibvirtMachineProviderSpec`

plus other identifiers that use `provider{c,C}onfig` string.

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -20,10 +20,11 @@ required = [
 
 [[constraint]]
   name = "sigs.k8s.io/cluster-api"
+  #revision = "0734939e05aeb64ab198e3feeee8b4e90ee5cbb2"
   branch = "master"
```
Comment (cynepco3hahue) [inline Gopkg.toml]: I think it better to use tag `0.0.0-alpha.4` because it includes all recent changes and will prevent broken API on `deps update`

Comment (ingvagabund) [reply Gopkg.toml]: That release does not contain https://github.com/kubernetes-sigs/cluster-api/pull/598. It's been unwritten rule to take a specific revision or master HEAD til now. Given we usually need the latest changes in the project, taking anything than master HEAD can be problematic.

Comment (cynepco3hahue) [reply Gopkg.toml]: I see your point but can we at least stick to some commit on the master?

Comment (ingvagabund) [reply Gopkg.toml]: If we do it here, we need to do it everywhere else. Though, we can wait until the next release is out and update all deps at once in all repos.

File: cmd/libvirt-actuator/main.go [hunk 2]
```diff
@@ -309,6 +309,12 @@ func bootstrapCommand() *cobra.Command {
 				glog.Info(msg)
 			}
 
+			clusterFramework.MachineControllerImage = "openshift/origin-libvirt-machine-controllers:v4.0.0"
+			clusterFramework.MachineManagerImage = "openshift/origin-libvirt-machine-controllers:v4.0.0"
+			// TODO(jchaloup): once docker.io/openshift/origin-machine-api-operator gets updated, switch back to it.
```
Comment (paulfantom) [inline cmd/libvirt-actuator/main.go]: It should get an update today, as https://github.com/openshift/machine-api-operator/pull/159 is on its way.

Comment (ingvagabund) [reply cmd/libvirt-actuator/main.go]: Who is responsible for making the updates?

Comment (paulfantom) [reply cmd/libvirt-actuator/main.go]: prow

**Comment by paulfantom** [comment]
/approve

putting it on hold since we should have a new mao image in docker hub today (this should do it: https://github.com/openshift/machine-api-operator/pull/159) and in turn  new image should resolve a TODO.

/hold

File: cmd/libvirt-actuator/main.go (hunk 2 above)
Comment (ingvagabund) [reply cmd/libvirt-actuator/main.go]: How often is it done?

Comment (paulfantom) [reply cmd/libvirt-actuator/main.go]: As far as I remember it should be done every couple of hours

Comment (ingvagabund) [reply cmd/libvirt-actuator/main.go]: The latest mao docker images is two weeks old :-|

Comment (paulfantom) [reply cmd/libvirt-actuator/main.go]: It seems like like CI image mirroring responsible for uploading images to docker hub was down between ~13 days ago and ~43 minutes ago.

File: Gopkg.toml (hunk 1 above)
Comment (ingvagabund) [reply Gopkg.toml]: Otherwise, the exact sha of the cluster-api is in the Godep.lock file.

File: Makefile [hunk 3]
```diff
@@ -83,7 +83,7 @@ build-e2e:
 
 .PHONY: test-e2e
 test-e2e: images build-e2e e2e-provision ## Run end-to-end test
-	hack/test-e2e.sh || $(MAKE) e2e-clean
+	hack/test-e2e.sh || ($(MAKE) e2e-clean && false)
```
Comment (frobware) [inline Makefile]: Why `&& false`?

File: pkg/cloud/libvirt/actuators/machine/actuator.go [hunk 4]
```diff
@@ -308,12 +308,22 @@ func (a *Actuator) deleteVolumeAndDomain(machine *clusterv1.Machine, client libv
 // ProviderConfigMachine gets the machine provider config MachineSetSpec from the
 // specified cluster-api MachineSpec.
 func ProviderConfigMachine(codec codec, ms *clusterv1.MachineSpec) (*providerconfigv1.LibvirtMachineProviderConfig, error) {
-	if ms.ProviderConfig.Value == nil {
+	// TODO(jchaloup): Remove providerConfig once all consumers migrate to providerSpec
+	providerSpec := ms.ProviderConfig
+	// providerSpec has higher priority over providerConfig
+	if ms.ProviderSpec.Value != nil || ms.ProviderSpec.ValueFrom != nil {
+		providerSpec = ms.ProviderSpec
+		fmt.Printf("Falling to default providerSpec\n")
```
Comment (frobware) [inline pkg/cloud/libvirt/actuators/machine/actuator.go]: Log instead of printing.

File: pkg/cloud/libvirt/actuators/machine/actuator.go [hunk 5]
```diff
@@ -308,12 +308,22 @@ func (a *Actuator) deleteVolumeAndDomain(machine *clusterv1.Machine, client libv
...
+	} else {
+		fmt.Printf("Falling to providerConfig\n")
```
Comment (frobware) [inline pkg/cloud/libvirt/actuators/machine/actuator.go]: Log instead of printing.

File: Makefile (hunk 3 above)
Comment (ingvagabund) [reply Makefile]: Since if `hack/test-e2e.sh` fails we need to clean the libvirt resource and fail

Comment (frobware) [reply Makefile]: Why doesn't `hack/test-e2e.sh` exit with non-zero if the test fails? If it doesn't we should fix the root cause.

Comment (frobware) [reply Makefile]: Oh, I see. If it fails,we clean up but we still want to return non-zero to make.

**Comment by ingvagabund** [comment]
/hold cancel
we can switch to the docker image at any point.

**Pull request merged**